### PR TITLE
Support one-time trips

### PR DESCRIPTION
**Pull request by binh-dam-ibigroup**
## Description

Works with https://github.com/ibi-group/otp-middleware/pull/224

This PR adds support for saving/monitoring trips that will be taken only once.

Behaviors of note:
- Checking the "Only on `<date>`" checkbox will uncheck other monitored days. Checking a monitored day will uncheck "Only on `<date>`".
- For one-time trips in the "trip is active" state, the state will switch to "Trip in the past" after the trip end time (with real-time adjustments if applicable).
- The requirement to have at least one monitored day between Monday and Sunday is removed.

## PR Checklist
- [x] Does the code follow accessibility standards (WCAG 2.1 AA Compliant)?
- [x] Are all languages supported (Internationalization/Localization)?
- [x] Are appropriate Typescript types implemented?

| Before | After |
|--------|-------|
| Before screenshot | after screenshot |

### CODE REVIEW

File: lib/components/user/monitored-trip/trip-basics-pane.tsx [hunk 1]
```diff
@@ -266,6 +269,12 @@ class TripBasicsPane extends Component<TripBasicsProps> {
                 )
               })}
             </AvailableDays>
+            <Checkbox checked={isOneTime} onChange={this._handleOneTimeTrip}>
+              <FormattedMessage
+                id="components.TripBasicsPane.onlyOnDate"
+                values={{ date: itinerary.startTime }}
+              />
+            </Checkbox>
```
Comment (miles-grant-ibigroup) [inline lib/components/user/monitored-trip/trip-basics-pane.tsx:277]: Should this maybe be a radio button instead? I'm thinking of having a big toggle switch between "monitor once" and "monitor regularly" and that way we can hide the calendar view if we have to

**Review by miles-grant-ibigroup — COMMENTED** [review]
Most of the new code looks good, just one small nit about the UI for the new feature

File: lib/components/user/monitored-trip/trip-basics-pane.tsx (hunk 1 above)
Comment (binh-dam-ibigroup) [reply lib/components/user/monitored-trip/trip-basics-pane.tsx]: Using radio buttons in https://github.com/opentripplanner/otp-react-redux/pull/1186/commits/29567d46154d705ed278aef20fb24748f3da24e2. I added the hide thing in https://github.com/opentripplanner/otp-react-redux/pull/1186/commits/d39120739f345b5586bec26377e9e4cafe17ab27. I think that makes things clearer.

**Review by miles-grant-ibigroup — APPROVED** [review]
Good enough for now!

**Review by daniel-heppner-ibigroup — APPROVED** [review]
This looks good, pending some kind of correction to the current behavior which breaks if the middleware is unable to validate a trip exists.

File: lib/components/user/monitored-trip/trip-status-rendering-strategies/past-trip-renderer.tsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+import baseRenderer from './base-renderer'
+
+/**
+ * Rendering info for a one-time trip in the past.
+ */
+export default function pastTripRenderer(
+  monitoredTrip: MonitoredTrip
+): MonitoredTripRenderData {
+  const data = baseRenderer(monitoredTrip) as unknown as MonitoredTripRenderData
```
[the remark this answers is not in the document]
Comment (binh-dam-ibigroup) [reply lib/components/user/monitored-trip/trip-status-rendering-strategies/past-trip-renderer.tsx:14]: If I just write `as MonitoredTripRenderData`, TypeScript doesn't believe me...

File: lib/components/user/monitored-trip/trip-basics-pane.tsx [hunk 3]
```diff
@@ -219,71 +258,88 @@ class TripBasicsPane extends Component<TripBasicsProps> {
...
-                      { repeatedDay: getFormattedDayOfWeekPlural(day, intl) }
+          <FormGroup>
+            <ControlLabel>
+              <FormattedMessage id="components.TripBasicsPane.tripDaysPrompt" />
+            </ControlLabel>
+            <Radio checked={!isOneTime} onChange={this._handleRecurringTrip}>
+              <FormattedMessage id="components.TripBasicsPane.recurringEachWeek" />
+            </Radio>
+            {!isOneTime ? (
+              <>
```
[the remark this answers is not in the document]
Comment (binh-dam-ibigroup) [reply lib/components/user/monitored-trip/trip-basics-pane.tsx:280]: > The `map` in there is also a lot to parse. Feels like a bit too much to tuck inline.

There is a redesign of this portion upcoming, so we will address that at that time.

**Comment by binh-dam-ibigroup** [comment]
> This looks good, pending some kind of correction to the current behavior which breaks if the middleware is unable to validate a trip exists.

@daniel-heppner-ibigroup Let me know what you think of this fix https://github.com/opentripplanner/otp-react-redux/pull/1186/commits/655454c7bf11d0315720adeab2633879d48d03af.

File: lib/components/user/monitored-trip/trip-basics-pane.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (binh-dam-ibigroup) [reply lib/components/user/monitored-trip/trip-basics-pane.tsx:299]: >  Could it just be `!isOneTime &&` in that case?

Refactored in https://github.com/opentripplanner/otp-react-redux/pull/1186/commits/655454c7bf11d0315720adeab2633879d48d03af.

File: lib/components/user/monitored-trip/trip-status-rendering-strategies/past-trip-renderer.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (daniel-heppner-ibigroup) [reply lib/components/user/monitored-trip/trip-status-rendering-strategies/past-trip-renderer.tsx:14]: Ah I see, the file it's in isn't typescripted.

**Review by daniel-heppner-ibigroup — APPROVED** [review]
Looks great! Thank you! @binh-dam-ibigroup

**Pull request merged**